feat: add auth, notifications, and messages to navbar

luchaos opened the pull request:
This PR is inspired by the prototype, the final one addressing the navbar whilst adding new elements and features that have been replaced in the top area:

**Remove top-card, add brand top with square logo, text and claim**

_Dark_
<img width="1562" alt="Screenshot 2023-09-05 at 17 27 46" src="https://github.com/RetroAchievements/RAWeb/assets/1280590/798c173f-577c-4e2e-862e-2c2711fe2738">
_Light_
<img width="1534" alt="Screenshot 2023-09-05 at 17 28 36" src="https://github.com/RetroAchievements/RAWeb/assets/1280590/08dbe0bc-a160-4a8d-993e-528f4a627359">

**Add bell icon to nav with badge for claim expiry & tickets notifications, add envelope icon to nav with badge for messages**

<img width="425" alt="Screenshot 2023-09-05 at 15 59 21" src="https://github.com/RetroAchievements/RAWeb/assets/1280590/22b14d29-13b6-40be-a711-ab5e1654cabf">

_Claim expiry & ticket notifications moved to the dropdown_
<img width="477" alt="Screenshot 2023-09-05 at 17 58 47" src="https://github.com/RetroAchievements/RAWeb/assets/1280590/81fa6526-f5c9-46eb-9cc0-112a6fcd632b">

_Small screens (375px+)_
<img width="354" alt="Screenshot 2023-09-05 at 16 52 35" src="https://github.com/RetroAchievements/RAWeb/assets/1280590/72156398-1f6b-4087-9601-bf0f08ca1c4f">

**Add "Sign in" & "Sign up" to navbar**

<img width="384" alt="Screenshot 2023-09-05 at 18 01 09" src="https://github.com/RetroAchievements/RAWeb/assets/1280590/ffb2a726-69f5-4115-9a9b-41002765e046">

**(Re)enable fortify as auth handler**:
- dedicated login route (/login) with prompt layout, replacing login request script
- login prompt inspired by GitHub's login & registration
- rate limiting on login attempts
- use existing form blade components
- "remember me" checkbox
- login event triggers writing activity (updates last login)

_Dark_
<img width="1051" alt="Screenshot 2023-09-05 at 18 34 03" src="https://github.com/RetroAchievements/RAWeb/assets/1280590/314c36b0-21ce-49cf-ab45-124d22d5f6f0">

_Light_
<img width="978" alt="Screenshot 2023-09-05 at 18 33 52" src="https://github.com/RetroAchievements/RAWeb/assets/1280590/41cfa437-b857-4e60-aec4-6010418a26b8">


**Registration page requires accepted terms**

<img width="484" alt="Screenshot 2023-09-05 at 15 53 59" src="https://github.com/RetroAchievements/RAWeb/assets/1280590/5c6a65bf-6efe-4677-b5a3-d073f01fe600">

Registration will be replaced with fortify eventually.

**Planned Follow-ups**
- Laravel Fortify: registration, email verification, password management, 2FA, ...
- Laravel Passport: prompt layout will be used for OpenID auth screens

Jamiras commented on resources/views/components/menu/notifications.blade.php:
```diff
@@ -1,19 +1,9 @@
+        </a>
+    </li>--}}
+    <livewire:notification-icon class="{{$class ?? ''}}" />
+    <livewire:message-icon class="{{$class ?? ''}}" />
```
I'm not sure if there's an easy solution for this, but I find myself trying to move diagonally from the icon to the menu item, which mouses over the Manage button and changes the menu before I can click on the notification submenu item.

![image](https://github.com/RetroAchievements/RAWeb/assets/32680403/10893a1e-b2ef-400d-9dcf-a9adbcf2f80e)

Jamiras replied:
There is much less right-padding when there are no notifications
![image](https://github.com/RetroAchievements/RAWeb/assets/32680403/db41c073-0d9a-41b4-b1e9-9af884d14713)

luchaos replied:
I found myself running into this frequently, too.
By removing the fixed width of 250px of dropdown items it would adjust its width automatically and there's maybe less of an urge to go sideways?
<img width="428" alt="Screenshot 2023-09-05 at 20 21 56" src="https://github.com/RetroAchievements/RAWeb/assets/1280590/2e09d0e7-23cb-46c7-994b-96cd1fefcfb0">
It would allow the games dropdown to scale more naturally, too, actually. It's very wide already.
<img width="914" alt="Screenshot 2023-09-05 at 20 23 50" src="https://github.com/RetroAchievements/RAWeb/assets/1280590/d32ab496-6d9c-45a2-8c49-f286ac9559fc">

luchaos replied:
Maybe right aligned just for the notifications dropdown? It doesn't work too well for the management and profile dropdowns, but feels better for notifications:
<img width="523" alt="Screenshot 2023-09-05 at 20 49 57" src="https://github.com/RetroAchievements/RAWeb/assets/1280590/d340cc99-bbef-42ce-bcd9-1e701de2dab5">

Jamiras replied:
I think the right alignment is enough for me to move the cursor straight down instead of down and left, as that's where the words are instead of a gap my brain says isn't part of the link.

Jamiras commented on resources/views/components/menu/brand.blade.php:
```diff
@@ -1,9 +1,3 @@
-        <img class="h-6" src="{{ asset('assets/images/ra-icon.webp') }}" alt="{{ config('app.name') }}">
-    @endif
+<a class="ml-3 {{ request()->routeIs('home') ? 'lg:hidden' : '' }}" href="{{ route('home') }}">
+    <img class="h-6" src="{{ asset('assets/images/ra-icon.webp') }}" alt="{{ config('app.name') }}">
```
I'm not sure how well this comes across in an image, but if you scroll down on the main page so the navbar is docked to the top, then go to some other page, the menu elements shift down a pixel (but not the red dots)

![image](https://github.com/RetroAchievements/RAWeb/assets/32680403/a62fdae4-e7d3-4cc0-8ad6-e1cbcbb37f5d)

I assume the icon is causing the height of the navbar to change a little.

luchaos replied:
What an interesting find! Turned out that rounding errors in the top area texts' line heights cause the height of the top area to shift ever so slighty, affecting the whole page.
Solved by setting a fixed height for the top brand area: https://github.com/RetroAchievements/RAWeb/pull/1831/commits/954dfefbcebd17b9d4149ccf4a77973098d6da15

Jamiras commented on resources/views/components/menu/account.blade.php:
```diff
@@ -75,8 +72,8 @@
-        <x-form :action="url('request/auth/logout.php')">
-            <button class="dropdown-item">{{ __('Sign Out') }}</button>
+        <x-form :action="route('logout')">
+            <button class="dropdown-item">{{ __('Sign out') }}</button>
```
message/icon.blade.php doesn't appear in this diff, so I'm putting the comment here indirectly.

Now that we have a Messages icon at the top level, do we still need a Messages menu item in the Account menu? Alternatively, should we remove the Messages button from the Nav Bar, keep the Messages link in the Account menu, and list unread messages under Notifications (with the tickets/claim notifications)?

If we keep the navbar item for notifications, would it be reasonable to report the number of unread messages in the tooltip for the navbar button? "Messages (2 unread)"?

luchaos replied:
I think adding "x unread messages" as a notification item is a neat idea.
It would cut down the amount of items listed there.

luchaos replied:
<img width="523" alt="Screenshot 2023-09-05 at 20 49 57" src="https://github.com/RetroAchievements/RAWeb/assets/1280590/9c730dea-53ff-4171-a925-c1f22f51ccb6">

See https://github.com/RetroAchievements/RAWeb/pull/1831/commits/358fe16e415f39a6d6e9ca2434cf3dde4f2e5e96

Jamiras commented on resources/views/auth/login.blade.php:
```diff
@@ -0,0 +1,22 @@
+                {{ __('Sign up') }}
+            </a>
+        </div>
+    </x-form>
```
Failed login reports the error twice.
![image](https://github.com/RetroAchievements/RAWeb/assets/32680403/3db5ff7c-f3e8-4298-8303-874260d441fd)

Jamiras replied:
Successful login always returns to the home page. Before this PR, the login would remain on the current page. I'm not sure if it's a deal-breaker, but I often notice that I'm not logged in when trying to access the Dev dropdown on a game page and seeing it's not there. Being able to login without losing my context was a nice feature. I understand that having an intermediate page complicates things.

wescopeland replied:
It would be ideal if there was some kind of callbackUrl or redirectUrl that after login is successful, sends the user back to that URL. Could be passed to the page as a query param.

luchaos replied:
Laravel supports that put of the box as "intended_url" session parameter and should redirect back automatically.
I'll look into why that's not happening.

luchaos replied:
Seems that redirect after login has to be done by hand with fortify 🤔 
Fixed in https://github.com/RetroAchievements/RAWeb/pull/1831/commits/704d4bc4e40ba259b41c9bad114c1796287b6be9

Removed global messages from prompt layout in https://github.com/RetroAchievements/RAWeb/pull/1831/commits/e27de4cbf0818b3ff02e6961616f5a33f26e6c22

Jamiras commented on resources/views/components/menu/management.blade.php:
```diff
@@ -5,81 +5,66 @@
-        @endcan--}}
-        @if($user->Permissions >= Permissions::Moderator)
-            <x-dropdown-item :link="url('admin.php')">Admin Tools</x-dropdown-item>
+    <div class="md:flex">
```
On sizes less than LG, the nav bar doesn't float (it scrolls with the page). Is that intentional?

luchaos replied:
Yes, kinda. It's still the same behaviour as before and the easiest way to make sure dropdowns with long lists of items can be scrolled down fully without being cut off at the bottom on small screens. 
There are certainly more sophisticated solutions for that (make the dropdown fixed width, make itself scrollable, make menus scrolalble overlays to make proper use of small screens, etc) but that might be a good candidate for an improvement PR.

Jamiras commented on resources/views/components/menu/account.blade.php:
```diff
@@ -7,24 +7,21 @@
-                <span class="hidden lg:inline-block">{{ __('Sign In') }}</span>
+    <div class="nav-item">
+        <a class="nav-link" href="{{ route('login') }}">
+            <x-fas-power-off class="mr-1" />
```
It feels odd for Sign In to have an icon, but not Sign Up.
![image](https://github.com/RetroAchievements/RAWeb/assets/32680403/ff74f906-8cb4-4cc8-9eb0-e33f50ded990)

Perhaps `fas-user-plus` for sign up, and `fas-user-pen` for sign in? 
![image](https://github.com/RetroAchievements/RAWeb/assets/32680403/0fcb7949-613d-4c71-8673-b3c07cb35a76)

luchaos replied:
Was thinking about that, too, but got close to removing the icon from "Sign in" rather
(I used the very same on completionist.me 😅 which is fine)
That looks very bare on the other hand 
<img width="355" alt="Screenshot 2023-09-05 at 21 13 26" src="https://github.com/RetroAchievements/RAWeb/assets/1280590/f43e48bc-4d13-40fd-9631-b62103d81871">

luchaos replied:
Updated in https://github.com/RetroAchievements/RAWeb/pull/1831/commits/0d90580495763a63bf61ef6ac7b376d928eec545
![Screenshot 2023-09-06 at 23 07 55](https://github.com/RetroAchievements/RAWeb/assets/1280590/9b4a774d-6f3c-4845-94de-2fe356360dc7)
![Screenshot 2023-09-06 at 23 08 06](https://github.com/RetroAchievements/RAWeb/assets/1280590/c4d68d87-a985-4edb-ab98-b3fe3ed60553)

luchaos commented:
I forgot to mention, please copy `storage/app/settings.dist.json` to `storage/app/settings.json` (see https://github.com/RetroAchievements/RAWeb/pull/1831/commits/781b0dee4a716cad10588487b7670abb0114feab).
This way it will add the system management links and you can add your own:
<img width="619" alt="Screenshot 2023-09-05 at 21 00 20" src="https://github.com/RetroAchievements/RAWeb/assets/1280590/6c3106f6-9032-4b6e-a9ad-cb8c3146cda8">

wescopeland commented:
nit: In lang/en_US/app.php:

```php
<?php

return [
    'description' => 'Adding achievements to your favourite retro games since 2012',
];
```

This is the text appearing in the heading, however "favourite" is not the correct en_US spelling. That is the spelling for en_GB. The correct spelling for en_US is "favorite".

wescopeland commented:
On mobile, there is a strange gap at the top of the page:

![Screenshot 2023-09-05 at 4 21 10 PM](https://github.com/RetroAchievements/RAWeb/assets/3984985/0af90a93-87d1-4385-bf68-b00a1da0adbc)

wescopeland commented:
Lighthouse is throwing an a11y warning on the notification button:

![Screenshot 2023-09-05 at 4 22 20 PM](https://github.com/RetroAchievements/RAWeb/assets/3984985/95a83c8f-1d8a-4ad0-9eb5-52b455c4fe01)

Jamiras commented:
> This is the text appearing in the heading, however "favourite" is not the correct en_US spelling. That is the spelling for en_GB. The correct spelling for en_US is "favorite".

That's the site slogan. Scott is British. It's been that way for quite some time.

![image](https://github.com/RetroAchievements/RAWeb/assets/32680403/2c5e1c64-8b23-4a8a-869b-bc6d32d60b5a)

But yes, it does feel wrong to have that in the en_US file.

wescopeland commented:
Yes, I am not disputing those facts. I am only suggesting that because we are going to surface this copy more heavily and because we are leaning into i18n and l10n more strongly, we should aim for correctness in our locale files.

luchaos commented:
that's just because en_us is the default, i had en_GB as default for some time but that seemed counterproductive. it wouldn't change for german or portugese. the slogan is technically correct like this i feel, but if scott is fine with it we can change it and add this spelling to en_GB

luchaos commented:
or drop "your favourite" altogether.

wescopeland commented:
nit: I really like how this page looks! One thing I am wondering, since both fields are required would it be appropriate to drop the "*" designator on the labels? If everything is required, it may be redundant to specify it as so. No strong opinion on this one, but wanted to throw it out there.

![Screenshot 2023-09-05 at 4 35 49 PM](https://github.com/RetroAchievements/RAWeb/assets/3984985/3e557094-f936-431d-86fd-9608a0140d4f)

wescopeland commented on resources/views/components/form-field.blade.php:
```diff
@@ -0,0 +1,39 @@
+            {{ $slot }}
+        @endif
+        @if(!empty($help))
+            <p class="help-block text-secondary mb-0">
+                {!! $help  !!}
+            </p>
+        @endif
+        @if($attribute ?? null)
+            @error($attribute)
+            <p class="help-block text-danger mb-0">
+                <x-fas-exclamation-triangle /> {{ $errors->first($attribute) }}
+            </p>
```
When these are visible, is there a good way for the input field to have an `aria-describedby` attribute that points to their IDs?

luchaos replied:
Phew, that's a neat detail. Addressed in https://github.com/RetroAchievements/RAWeb/pull/1831/commits/12f89de8f7c536e07e7de6f667e008c11be29367

wescopeland reviewed (commented):
Only one thing other than what I've already noted jumped out at me. Overall I think this is a great change!

luchaos commented:
> On mobile, there is a strange gap at the top of the home page:
> 
> ![Screenshot 2023-09-05 at 4 21 10 PM](https://user-images.githubusercontent.com/3984985/265817116-0af90a93-87d1-4385-bf68-b00a1da0adbc.png)

Whoops, good catch; fixed in https://github.com/RetroAchievements/RAWeb/pull/1831/commits/a9253fcff2613631fcc2693691de4a5e923f5115

luchaos commented:
> Lighthouse is throwing an a11y warning on the notification button:

Fixed in https://github.com/RetroAchievements/RAWeb/pull/1831/commits/bee5bddaf0aa82fbc1c6d446aa7821f179925991

I found that the open graph prefix was incorrectly set in the doctype which caused a lower lighthouse score due to falling back to quirks mode. Moved it to head, where the open graph meta tags are actually used, and the score increased for best practices, too: https://github.com/RetroAchievements/RAWeb/pull/1831/commits/788a75ab846c351c6726a7a571ee8bb464ee6272

luchaos commented:
Slogan update and alignment in https://github.com/RetroAchievements/RAWeb/pull/1831/commits/5720b123709036e3430ba7beace6dc36f1fa6b35
![Screenshot 2023-09-06 at 23 17 03](https://github.com/RetroAchievements/RAWeb/assets/1280590/fb02f908-5d94-404d-b4bc-cf3dc1894776)
